Cabal/src/Distribution/Simple/Utils.hs

```diff
@@ -372,7 +379,23 @@ die' verbosity msg = withFrozenCallStack $ do
     =<< annotateErrorString verbosity
     =<< pure . wrapTextVerbosity verbosity
     =<< pure . addErrorPrefix
-    =<< prefixWithProgName msg
+    =<< prefixWithProgName msg      
+
+-- Type which will be a wrapper for cabal -expections and cabal-install exceptions
+data VerboseException a  = VerboseException CallStack POSIXTime Verbosity a
+ deriving (Show, Typeable)
+
+-- A new dieWithException function which will replace the existing die' call sites
+dieWithException :: HasCallStack => Verbosity -> CabalException -> IO a
+dieWithException verbosity exception = do
+  ts <- getPOSIXTime
+  throwIO $ VerboseException callStack ts verbosity exception
+
+instance (Show a ,Typeable a) => Exception (VerboseException a) where
+_displayException :: VerboseException CabalException -> [Char]
+_displayException (VerboseException stack timestamp verb cabalexception) =  
```

Comment:
Perhaps, you should remove the `_` and add indentation, otherwise you're getting the default implementation of `displayException = show`, and that's how you get a `show` of an exception in the currently failing test on CI: 

After turning on your `displayException`, the code won't typecheck because you call `exceptionMessage cabalexception` which assumes that the instance have the form `Exception (VerboseException CabalException)` rather than the current polymorphic form (with an `a`). It's easy to fix: change the form of instance to the monomorphic one (you'll also need `FlexibleInstances`). 

After you fix this, you'll see again that the tests for error handling are no longer passing. You'll have to tweak the format of your output a bit on one hand, and update the tests to reflect the new bit of info -- the error codes. See https://github.com/haskell/cabal/blob/master/cabal-testsuite/README.md on how to `--accept` new test output. 

Generally, the whole idea was, as I understand it, to thread slowly and keep tests passing. This is currently not the case: the difference in the output is already visible.

Reply:
Forgot to add a link showing that you are currently using `displayException = show`: https://github.com/haskell/cabal/actions/runs/5357570778/jobs/9718610451?pr=9018#step:16:112

Reply:
This is still not addressed, I believe. You need to start getting the error codes from the modules that you've ported to structured errors. And for that you need to rename `_displayException` into `displayException`.

Reply:
I have fixed the `displayException` part and changed to monomorphic instance. While I committed this change along with the changes in the other module I remember the CI got passed. I missed updating here ,sorry.

Reply:
No worries at all! Looks good now. Let's see what CI says. I think some tests should fail because some error messages should change and produce the error codes now. ALthough, maybe our test coverage of error message may be worse than I'm imagining.